File: app/src/main/java/cn/ifafu/ifafu/mvp/elec_main/ElecMainContract.java
package cn.ifafu.ifafu.mvp.elec_main;

import java.util.Collection;
import java.util.List;
import java.util.Map;

import cn.ifafu.ifafu.data.entity.ElecQuery;
import cn.ifafu.ifafu.mvp.base.i.IModel;
import cn.ifafu.ifafu.mvp.base.i.IPresenter;
import cn.ifafu.ifafu.mvp.base.i.IView;
import cn.ifafu.ifafu.data.entity.Selection;
import io.reactivex.Observable;

public class ElecMainContract {

    interface View extends IView {

        void setElecText(String text);

        void setSelectorView(int i, List<String> list);

        void setBalanceText(String text);

        void showElecCheckView();

        void showPayView();

        void initViewVisibility();

        void showConfirmDialog(String message);

        /**
         * 设置电控选项
         *
         * @param strings 电控名字
         */
        void showDKSelection(Collection<String> strings);

        void setRoomText(String text);

        /**
         * 设置电费选项信息，用于快捷查询的view初始化设置
         */
        void setSelections(String dkName, String area, String building, String floor);

        /**
         * @return 地区名称
         */
        String getAreaText();

        /**
         * @return 楼栋名称
         */
        String getBuildingText();

        /**
         * @return 楼层名称
         */
        String getFloorText();

        /**
         * @return 充值金额
         */
        String getPriceText();

        /**
         * @return 宿舍号
         */
        String getRoomText();

        /**
         * @return 选择的电控名字
         */
        String getCheckedDKName();
    }

    public interface Presenter extends IPresenter {

        void onAreaSelect(String name);

        void onBuildingSelect(String name);

        void onFloorSelect(String name);

        void onDKSelect(String aid);

        /**
         * 查询余额
         */
        void queryCardBalance();

        /**
         * 查询电费
         */
        void queryElecBalance();

        /**
         * 充值前弹窗
         */
        void whetherPay();

        /**
         * 确认充值
         */
        void pay();

    }

    interface Model extends IModel {

        String getAccount();

        String getXfbAccount();
        /**
         * 初始化Cookie
         *
         * @return 是否重新登录
         */
        Observable<Boolean> initCookie();

        List<Selection> getSelectionFromJson();

        Observable<Double> queryBalance();

        /**
         * 获取所有电控信息
         *
         * @return key: Name(电控名字) value: Aid(电控Id)
         */
        Observable<Map<String, String>> queryDKInfos();

        /**
         * 查询电费
         */
        Observable<String> queryElectricity(ElecQuery data);

        Observable<String> elecPay(ElecQuery data, String price);

        /**
         * 保存查询选项，用于一键查询
         */
        void save(ElecQuery data);

        /**
         * @return 一键查询数据
         */
        ElecQuery getQueryData();
    }

}

File: app/src/main/java/cn/ifafu/ifafu/mvp/elec_main/ElecMainContract.kt
package cn.ifafu.ifafu.mvp.elec_main

import cn.ifafu.ifafu.base.i.IModel
import cn.ifafu.ifafu.base.i.IPresenter
import cn.ifafu.ifafu.base.i.IView
import cn.ifafu.ifafu.entity.ElecQuery
import cn.ifafu.ifafu.entity.Selection
import io.reactivex.Observable

class ElecMainContract {

    interface View : IView {

        /**
         * @return 地区名称
         */
        fun getAreaText(): String

        /**
         * @return 楼栋名称
         */
        fun getBuildingText(): String

        /**
         * @return 楼层名称
         */
        fun getFloorText(): String

        /**
         * @return 充值金额
         */
        fun getPriceText(): String

        fun getRoomText(): String

        /**
         * @return 选择的电控名字
         */
        fun getCheckedDKName(): String

        fun setRoomText(room: String)

        fun setElecText(text: String)

        fun setSelectorView(i: Int, list: List<String>)

        fun setBalanceText(text: String)

        fun showElecCheckView()

        fun showPayView()

        fun initViewVisibility()

        fun showConfirmDialog(message: String)

        /**
         * 设置电控选项
         *
         * @param strings 电控名字
         */
        fun showDKSelection(strings: Collection<String>)

        /**
         * 设置电费选项信息，用于快捷查询的view初始化设置
         */
        fun setSelections(dkName: String, area: String, building: String, floor: String)

        fun openLoginActivity();
    }

    interface Presenter : IPresenter {

        fun onAreaSelect(name: String)

        fun onBuildingSelect(name: String)

        fun onFloorSelect(name: String)

        fun onDKSelect(aid: String)

        /**
         * 查询余额
         */
        fun queryCardBalance()

        /**
         * 查询电费
         */
        fun queryElecBalance()

        /**
         * 充值前弹窗
         */
        fun whetherPay()

        /**
         * 确认充值
         */
        fun pay()

    }

    interface Model : IModel {

        fun getSelectionFromJson(): List<Selection>

        /**
         * @return 查询数据
         */
        fun getQueryData(): ElecQuery?

        /**
         * 初始化Cookie
         *
         * @return 是否重新登录
         */
        fun initCookie(): Observable<Boolean>

        fun queryBalance(): Observable<Double>

        /**
         * 获取所有电控信息
         *
         * @return key: Name(电控名字) value: Aid(电控Id)
         */
        fun queryDKInfos(): Observable<Map<String, String>>

        /**
         * 查询电费
         */
        fun queryElectricity(data: ElecQuery): Observable<String>

        fun elecPay(data: ElecQuery, price: String): Observable<String>

        /**
         * 保存查询选项，用于一键查询
         */
        fun save(data: ElecQuery)
    }

}
